File: packages/lynx-list/ios/SigxListUI.swift
import Foundation
import UIKit
import Lynx

/// Native UI for the `sigx-list` platform tag (#844) — `LynxUICollection`
/// (the engine's `<list>`) plus a frame-cheap bottom content inset.
///
/// Registered via the autolinker — `signalx-module.json`'s `ios.uiComponents`
/// produces `config.registerUI(SigxListUI.self, withName: "sigx-list")` in
/// the generated registry. The JS `<List>` opts in per element with
/// `custom-list-name="sigx-list"` (only when its `bottomInset` prop is set),
/// so every other `<list>` keeps the engine class.
///
/// `setBottomInset` mutates `collectionView.contentInset.bottom` DIRECTLY —
/// no CSS write, no starlight layout pass, and deliberately NO
/// invalidation context: cell layout attributes are content-coordinate, the
/// inset only moves the viewport, which is what makes it cheap enough for a
/// worklet-driven SharedValue to invoke per frame.
///
/// Composition points with the engine (all "call super, then apply delta" —
/// no engine logic copied):
///  - `updateFrame(_:withPadding:...)` overwrites `contentInset` from CSS
///    padding on every relayout (`LynxUICollection.m`) — re-add the extra.
///  - The engine scroller's `alignTo:"bottom"` targets the VIEWPORT bottom,
///    ignoring extra inset (`LynxCollectionScroller.m`: `targetOffset =
///    -offset + cellOffset - viewportH + cellH`, inset appears only in the
///    clamp) — so the chat re-pin would land under the keyboard. The
///    `scrollToPosition` override biases `offset` by `-extraBottomInset` for
///    bottom-aligned scrolls; the clamp is already inset-aware. (Android
///    needs no such bias — RecyclerView's scroller is padding-aware.)
// Class is NOT marked `@objc` — Swift forbids that on generic subclasses
// (`LynxUI<__covariant V>` is an ObjC lightweight generic). Member-level
// `@objc` / `@objc(name)` annotations still bridge; `LynxUIMethodProcessor`
// walks the superclass chain, so the engine's own list methods keep working.
public class SigxListUI: LynxUICollection {

    /// Extra bottom inset currently applied, in points (== CSS px on iOS).
    private var extraBottomInset: CGFloat = 0

    /// Status codes: `kUIMethodSuccess = 0`. Defined in the engine's ObjC
    /// headers; redefined locally (webview precedent) to avoid importing a
    /// private header just for one constant.
    private static let kUIMethodSuccess: Int32 = 0

    // MARK: - setBottomInset (per-frame UI method)

    /// `setBottomInset({ inset, pin })` — `inset` in CSS px, `pin` marks chat
    /// mode's stick-to-bottom so the newest item stays anchored above the
    /// inset (same-frame compensation, no `layoutcomplete` round trip).
    @objc public func setBottomInset(_ params: NSDictionary?,
                                     withResult callback: @escaping LynxUIMethodCallbackBlock) {
        let target = max(0, CGFloat((params?["inset"] as? NSNumber)?.doubleValue ?? 0))
        let pin = (params?["pin"] as? NSNumber)?.boolValue ?? false
        let delta = target - extraBottomInset
        guard delta != 0, let cv = self.view() else {
            callback(SigxListUI.kUIMethodSuccess, nil)
            return
        }
        // Pinned = at (or within 1pt of) max scroll BEFORE the change. Only a
        // pinned viewport is compensated: growing/shrinking the inset never
        // moves anchored content, so a reader scrolled up sees nothing shift.
        let maxBefore = maxContentOffsetY(cv)
        let wasPinned = pin && cv.contentOffset.y >= maxBefore - 1

        extraBottomInset = target
        // Delta-measured write: UIKit does not adjust contentOffset when the
        // bottom inset grows (adjustment only re-clamps when the offset falls
        // OUTSIDE the new range), but measure around the write anyway so any
        // OS-version surprise is folded into the compensation below.
        let offsetBefore = cv.contentOffset.y
        cv.contentInset.bottom += delta
        let osAdjustment = cv.contentOffset.y - offsetBefore

        if wasPinned {
            let maxAfter = maxContentOffsetY(cv)
            cv.setContentOffset(CGPoint(x: cv.contentOffset.x, y: maxAfter), animated: false)
        } else if osAdjustment != 0 {
            // Not pinned: anchored content must not move — undo any
            // auto-adjust the OS applied.
            cv.setContentOffset(CGPoint(x: cv.contentOffset.x, y: offsetBefore), animated: false)
        }
        callback(SigxListUI.kUIMethodSuccess, nil)
    }

    @objc(__lynx_ui_method_config__setBottomInset)
    dynamic public class func __lynxUIMethodConfigSetBottomInset() -> NSString {
        return "setBottomInset"
    }

    /// Max scrollable offset with the CURRENT insets — same formula as the
    /// engine scroller's clamp (`LynxCollectionScroller.m`).
    private func maxContentOffsetY(_ cv: UICollectionView) -> CGFloat {
        return max(-cv.contentInset.top,
                   cv.contentSize.height - cv.bounds.height + cv.contentInset.bottom)
    }

    // MARK: - Engine composition

    /// The engine overwrites `contentInset` from CSS padding on every
    /// relayout — stack the extra back on so the inset survives.
    public override func updateFrame(_ frame: CGRect,
                                     withPadding padding: UIEdgeInsets,
                                     border: UIEdgeInsets,
                                     margin: UIEdgeInsets,
                                     withLayoutAnimation with: Bool) {
        super.updateFrame(frame, withPadding: padding, border: border, margin: margin,
                          withLayoutAnimation: with)
        if extraBottomInset != 0 {
            self.view()?.contentInset.bottom += extraBottomInset
        }
    }

    /// Bias bottom-aligned scrolls by the extra inset (see the class doc).
    /// Same selector as the engine's category method (`LYNX_UI_METHOD
    /// (scrollToPosition)` in `LynxUICollection+Delegate.m`) — the method
    /// dispatch resolves dynamically, so this runtime override wins without
    /// re-declaring a `__lynx_ui_method_config__`; super is reached via an
    /// IMP cast because the category method has no public declaration.
    @objc(scrollToPosition:withResult:)
    public func scrollToPosition(_ params: NSDictionary?,
                                 withResult callback: @escaping LynxUIMethodCallbackBlock) {
        var forwarded = params
        if extraBottomInset != 0,
           (params?["alignTo"] as? String) == "bottom" {
            let mutable = NSMutableDictionary(dictionary: params ?? [:])
            let offset = (params?["offset"] as? NSNumber)?.doubleValue ?? 0
            mutable["offset"] = NSNumber(value: offset - Double(extraBottomInset))
            forwarded = mutable
        }
        let sel = NSSelectorFromString("scrollToPosition:withResult:")
        guard let superclass = class_getSuperclass(object_getClass(self)),
              let method = class_getInstanceMethod(superclass, sel) else {
            callback(SigxListUI.kUIMethodSuccess, nil)
            return
        }
        typealias ScrollFn = @convention(c) (
            AnyObject, Selector, NSDictionary?, @escaping LynxUIMethodCallbackBlock
        ) -> Void
        let imp = method_getImplementation(method)
        unsafeBitCast(imp, to: ScrollFn.self)(self, sel, forwarded, callback)
    }
}

File: packages/lynx-list/android/com/sigx/list/SigxListUI.kt
package com.sigx.list

import android.util.Log
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.ReadableMap
import com.lynx.tasm.behavior.LynxContext
import com.lynx.tasm.behavior.LynxUIMethod
import com.lynx.tasm.behavior.LynxUIMethodConstants
import com.lynx.tasm.behavior.ui.list.UIList
import com.lynx.tasm.utils.PixelUtils
import kotlin.math.roundToInt

/**
 * `<list>` subclass with a frame-cheap bottom content inset (#844).
 *
 * `setBottomInset` adjusts the RecyclerView's bottom padding DIRECTLY —
 * no CSS write, no starlight layout pass — so a main-thread SharedValue
 * (keyboard lift, bottom-sheet drag) can drive it per frame via
 * `useAnimatedMethod`. `clipToPadding` is already false (`UIList.createView`),
 * so the padding acts as a pure viewport inset: growing it extends the
 * scrollable range without moving anchored content.
 *
 * Chat-pin composition: when the caller flags `pin` and the viewport sits at
 * the bottom BEFORE the change, the scroll position shifts by the inset delta
 * in the same call — the newest item stays anchored above the inset with no
 * `layoutcomplete` round trip. `setPadding` updates the padding fields
 * synchronously and the LayoutManager reads them live (`getEndAfterPadding`),
 * so the `scrollBy` consumes fully in the same frame.
 *
 * The engine re-applies CSS padding + border on every relayout
 * (`UIList.onLayoutUpdated` → `setPadding`), which would clobber the extra
 * inset — the override below stacks it back on, so the inset survives
 * relayouts. All engine access stays on public surface (`getRecyclerView`,
 * `onLayoutUpdated`, `RecyclerView.setPadding`/`scrollBy`); nothing is
 * copied from the engine, keeping the subclass resilient across engine bumps.
 */
class SigxListUI : UIList {
    constructor(context: LynxContext) : super(context)
    constructor(context: LynxContext, param: Any?) : super(context, param)

    init {
        // Verification breadcrumb for the custom-list-name opt-in: if this
        // never logs while a `bottomInset` list is on screen, the platform
        // tag didn't resolve to `sigx-list` (see #844).
        if (DEBUG) Log.d(TAG, "SigxListUI created for tag sigx-list")
    }

    /** Extra bottom inset currently applied, in device px (rounded). */
    private var extraBottomPx = 0

    /**
     * `setBottomInset({ inset, pin })` — `inset` in CSS px (dp), `pin` marks
     * chat mode's stick-to-bottom so the newest item stays anchored.
     * Cheap enough to invoke per frame from a worklet binding.
     */
    @LynxUIMethod
    fun setBottomInset(params: ReadableMap?, callback: Callback?) {
        val rv = recyclerView
        if (rv == null) {
            callback?.invoke(LynxUIMethodConstants.UNKNOWN)
            return
        }
        // Horizontal lists: vertical bottom inset is out of scope (v1) — the
        // JS side only wires vertical lists; succeed as a no-op for safety.
        if (rv.layoutManager?.canScrollVertically() != true) {
            callback?.invoke(LynxUIMethodConstants.SUCCESS)
            return
        }
        val insetCssPx = if (params != null && params.hasKey("inset")) params.getDouble("inset") else 0.0
        val pin = params != null && params.hasKey("pin") && params.getBoolean("pin")

        val newExtra = PixelUtils.dipToPx(insetCssPx.toFloat()).roundToInt().coerceAtLeast(0)
        val delta = newExtra - extraBottomPx
        if (delta == 0) {
            callback?.invoke(LynxUIMethodConstants.SUCCESS)
            return
        }
        // Pinned = at (or within 1px of) the bottom BEFORE the change. Only a
        // pinned viewport gets compensated: an inset change never moves
        // anchored content, so a user reading history sees nothing shift.
        val distanceFromBottom = rv.computeVerticalScrollRange() -
            rv.computeVerticalScrollExtent() - rv.computeVerticalScrollOffset()
        val wasPinned = pin && distanceFromBottom <= 1

        extraBottomPx = newExtra
        rv.setPadding(rv.paddingLeft, rv.paddingTop, rv.paddingRight, rv.paddingBottom + delta)
        if (wasPinned) {
            // Works for both signs: grow slides content up with the occluder,
            // shrink follows it back down.
            rv.scrollBy(0, delta)
        }
        callback?.invoke(LynxUIMethodConstants.SUCCESS)
    }

    /**
     * CSS relayout re-applies base padding + border (`UIList.onLayoutUpdated`
     * → `RecyclerView.setPadding`) — stack the extra inset back on so it
     * survives every relayout.
     */
    override fun onLayoutUpdated() {
        super.onLayoutUpdated()
        if (extraBottomPx != 0) {
            val rv = recyclerView ?: return
            rv.setPadding(rv.paddingLeft, rv.paddingTop, rv.paddingRight, rv.paddingBottom + extraBottomPx)
        }
    }

    companion object {
        private const val TAG = "SigxListUI"

        /** Flip to true to log instance creation (tag-resolution verification). */
        private const val DEBUG = false
    }
}
